server/src/main/java/io/druid/segment/realtime/plumber/CoordinatorBasedSegmentHandoffNotifier.java

```diff
@@ -0,0 +1,166 @@
+import com.metamx.common.logger.Logger;
+import io.druid.client.SegmentLoadInfo;
+import io.druid.client.coordinator.CoordinatorClient;
+import io.druid.concurrent.Execs;
+import io.druid.query.SegmentDescriptor;
+import io.druid.server.coordination.DruidServerMetadata;
+
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+public class CoordinatorBasedSegmentHandoffNotifier implements SegmentHandoffNotifier
+{
+  private static final Logger log = new Logger(CoordinatorBasedSegmentHandoffNotifier.class);
+
+  private final Map<SegmentDescriptor, Pair<Executor, Runnable>> handOffCallbacks = Maps.newConcurrentMap();
+  private final CoordinatorClient coordinatorClient;
+  private volatile ScheduledExecutorService scheduledExecutor;
+  private final long pollDurationMillis;
+  private final String dataSource;
+
+  public CoordinatorBasedSegmentHandoffNotifier(
+      String dataSource,
+      CoordinatorClient coordinatorClient,
+      CoordinatorBasedSegmentHandoffNotifierConfig config
+  )
+  {
+    this.dataSource = dataSource;
+    this.coordinatorClient = coordinatorClient;
+    this.pollDurationMillis = config.getPollDuration().getMillis();
+  }
+
+  @Override
+  public void registerSegmentHandoffCallback(
+      SegmentDescriptor descriptor, Executor exec, Runnable handOffRunnable
+  )
+  {
+    log.info("Adding SegmentHandoffCallback for dataSource[%s] Segment[%s]", dataSource, descriptor);
+    handOffCallbacks.put(descriptor, new Pair<>(exec, handOffRunnable));
+  }
+
+  @Override
+  public void start()
+  {
+    scheduledExecutor = Execs.scheduledSingleThreaded("coordinator_handoff_scheduled_%d");
+    scheduledExecutor.scheduleAtFixedRate(
+        new Runnable()
+        {
+          @Override
+          public void run()
+          {
+            try {
+              Iterator<Map.Entry<SegmentDescriptor, Pair<Executor, Runnable>>> itr = handOffCallbacks.entrySet()
+                                                                                                     .iterator();
+              while (itr.hasNext()) {
+                Map.Entry<SegmentDescriptor, Pair<Executor, Runnable>> entry = itr.next();
+                SegmentDescriptor descriptor = entry.getKey();
+                try {
+                  List<SegmentLoadInfo> loadedSegments = coordinatorClient.fetchServerView(
+                      dataSource,
+                      descriptor.getInterval(),
+                      true
+                  );
+
+                  if (isHandOffComplete(loadedSegments, entry.getKey())) {
+                    log.info("Segment Handoff complete for dataSource[%s] Segment[%s]", dataSource, descriptor);
+                    entry.getValue().lhs.execute(entry.getValue().rhs);
+                    itr.remove();
+                  } else {
+                    log.info(
```

Comment:
Print these all at the end of the loop?

Like: `Still waiting for handoff of segments: blah`